src/Extension/WorkflowFieldCapture.php

```diff
@@ -12,14 +12,18 @@ class WorkflowFieldCapture extends DataExtension
 {
     public function onBeforeWrite()
     {
+        // TODO: make this check for instance
         if ($this->owner->Comment) {
             // see if we've got an active workflow that might be interested in the
             // comment text
             $active = $this->owner->getWorkflowInstance();
             if ($active) {
                 $action = $active->CurrentAction();
                 if ($action) {
-                    $action->update($this->owner->toMap());
+                    $postVars = Controller::curr()->getRequest()->postVars();
```

Comment:
Does the default behaviour here still do what the previous behaviour did, whereby _all_ fields that come through the submit get written to the object?

Reply:
It's more that this wasn't doing what was intended to begin with I think. The previous code is writing a **Page**'s fields into the **Action** object/table. Thus:

- The update() uses the Page's ID to find an action, which either:
  - Fails and a new Action with the Page's ID is created.
  - Or it finds an action made as above.
  - Or it finds a valid action who's ID happens to be the same as the Page (and which will now be overwritten, invalidating it).
- The update() then writes any Page fields that overlap with the Action (ID, Created, etc). If you look at the WorkflowActionInstance table, you'll see lots of entries missing their ClassName, those are the failed attempts to write a Page into that table.
- IIRC Comment had, by that point, been wrongly assigned to the Page object, and so it was present in the `toMap()` call.
- Comment is now written into the action via `WorkflowActionInstanceExtension::onSaveWorkflowPage()`
- It's worth noting that no action was overriding the update() method.

It's very possible I've missed something in my implementation of all this, but either way the previous approach needs fixing.

Reply:
Yeah, that does look busted; I believe the original functionality was done that way because when a user hits 'save' or the _next_ workflow step button, anything entered in the right hand form needs to get saved into the workflow action instance (ie the Comment field, or any other field that action expects). Likely what needs to happen is that the `WorkflowActionInstanceExtension` needs to look at all the DB fields of the action instance _except a reserved list_ and set those values if they're present on the request. IIRC there's no way of accessing the form for the request by the time the actions are being triggered, so the only way to access the data (without grabbing the controller directly) is off the page being modified, which _does_ have the workflow action fields set against it directly on save. 

In its current state, this will break existing workflows that have additional properties being collected in the right column.

Reply:
I think we should discuss this IRL for clarity and speed.

Reply:
Solution discussed IRL now implemented. Key point is that we're now using `invokeWithExtensions()`.